EIPS/eip-FOCIL.md

```diff
@@ -26,10 +26,79 @@ FOCIL is a simple committee-based design improving upon previous IL mechanisms o
+| `MAX_GAS_PER_INCLUSION_LIST` | `2**21` |
+| `MAX_TRANSACTIONS_PER_INCLUSION_LIST_AGGREGATE` | `2 * MAX_TRANSACTIONS_PER_INCLUSION_LIST = 32` |
+| `MAX_GAS_PER_INCLUSION_LIST_AGGREGATE` | `2 * MAX_GAS_PER_INCLUSION_LIST = 2**22` |
+| `IL_COMMITTEE_SIZE` | `256` |
+
+### Reference Objects
+
+```
+class InclusionListEntry(Container):
+    from_address: ExecutionAddress
+    gas_limit: uint64
+```
+
+```
+class inclusionList(Container)
+    slot: Slot
+    proposer_index: ValidatorIndex
+    entries: List[InclusionListEntry, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
+```
+
+```
+class InclusionListAggregatedEntry(Container):
+    from_address: ExecutionAddress
+    gas_limit: uint64
+    bitlist: List[byte, IL_COMMITTEE_SIZE]
+```
+
+```
+class InclusionListAggregated(Container):
+    slot: Slot
+    proposer_index: ValidatorIndex
+    message: List[InclusionListAggregatedEntry, MAX_TRANSACTIONS_PER_INCLUSION_LIST_AGGREGATE]
+```
+
 ### Consensus layer
 
 #### High-level overview
 
+TBA (Same as timeline)
+
+#### Specific changes
+
+**Beacon chain state transition spec:**
+
+- ***New** `inclusion_list` object:* Introduce a new `inclusion_list` for the IL committee to submit and nodes to process.
+- ***Modified** `BeaconBlockBody`:* Modified to include the aggregated inclusion list for that slot.
+- ***Modified** `process_execution_payload` function:* Update this process to include checks for the inclusion list evaluation.
+
+**Beacon chain P2P spec:**
+
+- ***New** gossipnet and validation rules for inclusion list:* Define new rules for handling the inclusion list in the gossip network and validation.
+- ***New** RPC request and response network for inclusion list:* Establish a new network for sending and receiving inclusion lists.
+
+**Validator spec:**
+
+- ***New** duty for preparing `inclusion_list`:* Inclusion list committee members to prepate and sign their respective local inclusion list.
+- ***New** proposer duty for aggregating `inclusion_list`:* Proposer to prepare an aggregated inclusion list and sign it.
+- ***New** attester duty for aggregating `inclusion_list`:* Attesters to prepare an aggregared inclusion list based on messages received for evaluation.
+- ***Modified** duty for `BeaconBlockBody`:* Update the duty to prepare the beacon block body containing `inclusion_list_aggregated` and satisfying transaction entries if block is not full.
+
+### Execution layer
+
+- ***New** `get_inclusion_list`:* Introduce a new function for the IL committee to retreive inclusion lists.
+- ***New** `new_inclusion_list`:* Define a new function for nodes to validate the execution side of the inclusion list (TBD).
+- ***Modified** `forkchoice_updated`:* Update the function with a `payload_attribute` to include the aggregated inclusion list as part of the attribute.
+- ***Modified** `new_payload`:* Update the function for EL clients to verify that `payload_transactions` satisfy `payload.inclusion_list_aggregated`.
```

Comment:
I am not sure if these 2 are correct because above we've mentioned that aggregated IL will be a part of the beacon block body so I am not sure if they need to be part of the execution payload or not. 

My question here is that where do a node validate if the payload satisfies the aggregated IL? I am assuming it would be in CL but because it might involve some nuanced checks about transactions, I thought it may also involve EL in some cases.

Reply:
by a node you mean an attester? if that's the case then yes some execution will be necessary there to ensure the validity of the payload (that includes IL_agg transactions)

Reply:
Yes I meant the attester. So, I am assuming the workflow would be something like this (correct me if I am wrong). 

CL
- The attester receives a new payload containing IL_agg (which is included by the builder)
- The attester validates beacon block fields
- The attester validates if the IL_agg associated with the block matches with the one sent by that slot's proposer
- The attester evaluates the quality of IL_agg (proposer) by comparing it with it's self built IL_agg
- The attester sends the payload to EL

EL
- On receiving payload, validate everything what happens now (block, tx execution, etc)
- Validate if the set of transactions satisfy the transactions from IL_agg in the block (this is a newly added check)
- Update fork choice only if both the above things are valid

Reply:
> I think the third bullet point on the CL is unnecessary
Shouldn't the attester validate if the builder included the same IL which was proposed by the slot proposer earlier?